planner/funcdep/fd_graph.go

```diff
@@ -549,6 +549,280 @@ func (s *FDSet) MakeCartesianProduct(rhs *FDSet) {
+//      - If the right side has rows, no row is filtered out after the filters since no row of the outer side is filtered out. Hence, the equivalence class remains.
+//
+func (s *FDSet) MakeOuterJoin(innerFDs, filterFDs *FDSet, outerCols, innerCols FastIntSet, opt *ArgOpts) {
+	//  copy down the left PK and right PK before the s has changed for later usage.
+	leftPK, ok1 := s.FindPrimaryKey()
+	rightPK, ok2 := innerFDs.FindPrimaryKey()
+	copyLeftFDSet := &FDSet{}
+	copyLeftFDSet.AddFrom(s)
+	copyRightFDSet := &FDSet{}
+	copyRightFDSet.AddFrom(innerFDs)
+
+	for _, edge := range innerFDs.fdEdges {
+		// Rule #2.2, constant FD are removed from right side of left join.
+		if edge.isConstant() {
+			continue
+		}
+		// Rule #2.3, equivalence FD are removed from right side of left join.
+		if edge.equiv {
+			continue
+		}
+		// Rule #2.1, lax FD can be kept after the left join.
+		if !edge.strict {
+			s.addFunctionalDependency(edge.from, edge.to, edge.strict, edge.equiv)
+			continue
+		}
+		// Rule #2.1, strict FD can be kept when determinant contains not null column, otherwise, downgraded to the lax one.
+		//
+		// If the one of the column from the inner child's functional dependency's left side is not null, this FD can be remained.
+		// This is because that the outer join would generate null-extended rows. So if at least one row from the left side
+		// is not null. We can guarantee that the there's no same part between the original rows and the generated rows.
+		// So the null extended rows would not break the original functional dependency.
+		if edge.from.Intersects(innerFDs.NotNullCols) {
+			// One of determinant are not null column, strict FD are kept.
+			// According knowledge #2, we can't take use of right filter's not null attribute.
+			s.addFunctionalDependency(edge.from, edge.to, edge.strict, edge.equiv)
+		} else {
+			// Otherwise, the strict FD are downgraded to a lax one.
+			s.addFunctionalDependency(edge.from, edge.to, false, edge.equiv)
+		}
+	}
+	leftCombinedFDFrom := NewFastIntSet()
+	leftCombinedFDTo := NewFastIntSet()
+	for _, edge := range filterFDs.fdEdges {
+		// Rule #3.2, constant FD are removed from right side of left join.
+		if edge.isConstant() {
+			s.Rule333Equiv.Edges = append(s.Rule333Equiv.Edges, &fdEdge{
+				from:   edge.from,
+				to:     edge.to,
+				strict: edge.strict,
+				equiv:  edge.equiv,
+			})
+			s.Rule333Equiv.InnerCols = innerCols
+			continue
+		}
+		// Rule #3.3, we only keep the lax FD from right side pointing the left side.
+		if edge.equiv {
+			// equivalence: {superset} --> {superset}, either `from` or `to` side is ok here.
+			// Rule 3.3.1
+			if !opt.SkipFDRule331 {
+				leftCover := edge.from.Intersection(outerCols)
+				rightCover := edge.from.Intersection(innerCols)
+				if leftCover.Len() > 0 && rightCover.Len() > 0 {
+					if opt.TypeFDRule331 == CombinedFD {
```

Comment:
Why do we need to distinguish between `CombinedFD` and `SingleFD`? I think regarding all cases as `CombinedFD` is OK.

Reply:
Sure, let check how to merge them with a clear comment and codes.